parsers/UY.py

```diff
@@ -46,118 +62,182 @@ def get_salto_grande(session: Optional[Session]) -> float:
-    )
-
-    obj = {"datetime": date.datetime}
-
-    for tr in table.find_all("tr"):
-        tds = tr.find_all("td")
-        if not len(tds):
-            continue
-
-        key = tds[0].find_all("b")
-        # Go back one level up if the b tag is not there
-        if not len(key):
-            key = tds[0].find_all("font")
-        k = key[0].contents[0]
-
-        value = tds[1].find_all("b")
-        # Go back one level up if the b tag is not there
-        if not len(value):
-            value = tds[1].find_all("font")
-        v_str = value[0].contents[0]
-        if v_str.find(",") > -1 and v_str.find(".") > -1:
-            # there can be values like "1.012,5"
-            v_str = v_str.replace(".", "")
-            v_str = v_str.replace(",", ".")
-        else:
-            # just replace decimal separator, like "125,2"
-            v_str = v_str.replace(",", ".")
-        v = float(v_str)
-
-        # solar reports -0.1 at night, make it at least 0
-        v = max(v, 0)
-
-        obj[k] = v
+    The function will first fetch data for all avaliable times,
+    then use make_output to format the data points before returning
+    a 24 hr history of data entries
+    """
+    # handle datetime here, so we only have to do it once
+    # get data from webpage
+    raw_entries = parse_page(session)
 
-    # https://github.com/tmrowco/electricitymap/issues/1325#issuecomment-380453296
-    salto_grande = get_salto_grande(session)
-    obj["Hidráulica"] = obj.get("Hidráulica", 0.0) + salto_grande
+    entries = []
+    for raw_entry in raw_entries:
+        if correct_hydro:
+            # https://github.com/tmrowco/electricitymap/issues/1325#issuecomment-380453296
+            raw_entry = correct_for_salto_grande(raw_entry, session)
+        entries.append(make_output(raw_entry))
 
-    return obj
+    return entries
+
+
+def fetch_consumption(
+    zone_key: str = "UY",
+    session: Optional[Session] = Session(),
+    target_datetime: Optional[datetime] = None,
+    logger: Logger = getLogger(__name__),
+) -> Union[dict, List[dict]]:
+
+    if target_datetime is not None:
+        raise ParserException("This parser is unnable to parse dates in the past")
+
+    # helper func to format data output
+    def make_datum(entry):
+        return {
+            "zoneKey": zone_key,
+            "datetime": entry["time"].datetime,
+            "consumption": entry["demand"],
+            "source": "ute.com.uy",
+        }
+
+    # then delegate all the actual work to get_entry_list
+    return get_entry_list(session, make_datum, correct_hydro=False)
 
 
 def fetch_production(
     zone_key: str = "UY",
-    session: Optional[Session] = None,
+    session: Optional[Session] = Session(),
     target_datetime: Optional[datetime] = None,
     logger: Logger = getLogger(__name__),
-) -> dict:
-    if target_datetime:
-        raise NotImplementedError("This parser is not yet able to parse past dates")
+) -> Union[dict, List[dict]]:
```

Comment:
The return type should only be `List[dict]` as we should always return lists now.

Reply:
Should be fixed in 950e3dad8ab32c0cea9d86c7fee46a794ccc36f5